### ISSUE TITLE
New design improvement spec: refinement of keyboard shortcuts

### ISSUE THREAD

**Issue by bkunda** [musescore/MuseScore#8746]
This task proposes updating the entire list of keyboard shortcuts in MS4 to adopt a new naming convention. 

The convention would adopt the format "verb"+"action".

- Instructions are in ~~[this spreadsheet](https://docs.google.com/spreadsheets/d/1nCiNP9YdKsCEfoK5K3Nm8jBst3Jn4cc4PFUE547dL0A/edit?usp=sharing)~~ [this spreadsheet](https://docs.google.com/spreadsheets/d/1IG0AgwEDcwMyXqs93K4_tSiyxvOtBs-znRTt3PkDGz4/edit?usp=sharing)

Thank you for your valuable assistance!

**Comment by cbjeukendrup** [comment]
I am working on this. Here is already the updated "View" menu. It's a little bit different from the spreadsheet, but I hope you agree on this. 
<img width="530" alt="Schermafbeelding 2021-08-14 om 15 01 39" src="https://user-images.githubusercontent.com/48658420/129447121-b7035c10-3841-4ff5-a666-1e8c555ede34.png">

**Comment by wizofaus** [comment]
BTW how do you intend to show the difference between multiple "key sequences" and "multiple key" sequences?
E.g. my shortcuts for "add bass clef" are "Ctrl+Y,Ctrl+2" and "Ctrl+;".  In MU3 this shows as "Ctrl+Y,Ctrl+2; Ctrl+;".
This isn't ideal as both ";" and "," can be part of a shortcut sequence so you can end up with weird things like "Ctrl+,,;; Ctrl+;"
(which means Ctrl + comma, then semi-colon, or Ctrl+semi-colon!).
But it's even less obvious in MU4 (it appears to be use comma+space as a separator for each sequence, and just comma as a separator for each key within a sequence, and worse, you can't even define multi-key sequences in Mu4, but some of the factory default sequences are like this, e.g. for Ottava 8va).

**Comment by Tantacrul** [comment]
Hmmm. Interesting. Because shortcut sequences are so completely broken in MS3 (on my Mac), I never realised this was how they were meant to work. Of course the UI is also incredibly unclear, so I'm not surprised our devs missed it - especially since 'Add' works inconsistently too. I wonder what fraction of users actually understand how this is meant to work?

Seems like a simple thing to accommodate though. We should just exactly match MS3 functionality for now (and clean up inconsistencies and non-functionality) and later we'll need to completely reimagine this system.

**Comment by Tantacrul** [comment]
@wizofaus - nice spot, btw!

**Comment by wizofaus** [comment]
There's also a bunch more commands here: https://github.com/musescore/MuseScore/issues/8819

I'd actually like a chance to discuss with the core team how to best implement some of these if possible, as the way things have been done currently is going to mean quite a lot more code than it should really need to be.

@Tantacrul

**Comment by wizofaus** [comment]
> 
> For now, we should just exactly match MS3 functionality (and clean up inconsistencies and non-functionality) and later we'll need to completely reimagine this interface and extend it to cover all functionality.

But MU3 is broken at least for me because half (4 out of 8) of the factory default multi-key sequences start with Ctrl+Y, and that already means "redo". I proposed making this Alt+Y for Windows at least (or better, coming up with a single key combination).

There's also 4 Shift+R,Shift+<another letter> sequences too (for adding rests of various lengths), including, confusingly "shift+r" twice for "add quarter note rest", whereas "shift+r, shift+q" adds an *1/8* note rest!  Would think it would make more sense to be, e.g. shift+1 followed by the number matching the duration (5 for 1/4 note, 4 for 1/8 note etc.).  But at least those ones work.

**Comment by wizofaus** [comment]
I notice also that currently in MU4:
a) It doesn't even *let* you re-use the same shortcut sequence for two different actions. But while for the most part this probably a good thing, it isn't for, e.g. change duration/fret actions. I'd accidentally cleared the shortcut sequence for "64th note" (due to below bug), which is normally just 1, but you can't reassign it as 1 is already assigned to "Fret 1 (TAB)". Reset to default works though.
b) when you have a search filter, the list doesn't update correctly after modifying/clearing/resetting the shortcut - in fact it modifies a different action's shortcut sequence entirely. This is pretty badly broken.
c) for plug-in development and trouble-shooting purposes it would be very handy to be able to see the name of the underlying command here (and ideally search for it!).  In some cases the descriptive names don't match the underlying command names well at all.

**Comment by wizofaus** [comment]
Also, I'm in the process of porting literally scores more (probably a 100!) actions from MU3 into MU4.
Currently there's no real concept of "categories" of actions, other than some that only apply in very particular context (e.g. in the score notation window, vs the palettes, vs at a global level).
Adding this is non-trivial and further, we need a list that defines what category each action should belong to.
Personally I think having "categories" in this dialog is very much a "nice-to-have" thing that could come later, but if we are going to prepare for it it makes sense to do the work now as part of what I'm doing porting the remaining actions.

**Comment by Tantacrul** [comment]
The categories thing is certainly something Bradely has already worked on for a later release of MS4. It would be super handy.

**Comment by bkunda** [comment]
@cbjeukendrup is this task still on your radar? I know you started working on it back in August. 
I'm thinking that just the first part of this task might be good to tackle (_just_ adopting the new naming convention – no change to any functionality), provided it doesn't mess anything else up ahead of the release. 
If you're still available for this, then perhaps we two can meet briefly to discuss it further.

**Comment by cbjeukendrup** [comment]
@bkunda I had indeed worked on this issue, but for some reason stopped (probably got distracted by other issues back then). I would be happy to pick this up again. 

Some changes I had made to the View menu are maybe not applicable anymore, since some other changes have been made meanwhile - mostly for the simple reason that the menu was too tall on some screens. But I believe we have a good solution for tall menus nowadays, so I could bring some options back.

**Comment by bkunda** [comment]
Thanks @cbjeukendrup! I see you've been chatting with @MarcSabatella over in Discord about a solution for pulling the command names from a central place and having them appear in the handbook. This would be a terrific help if you can assist with this :-) 

And just on changes to menu items, @MarcSabatella has made the point to me that each command may have two (possibly three?) different names, so that they appear differently in the shortcuts dialog than in menu items, etc. I'm not sure if this is still the case in MS4, but it is pretty important – the changes I've proposed should _only_ affect the list in Preferences > Shortcuts. Menu items have a different UX logic, so I wouldn't want to interfere with this

**Comment by cbjeukendrup** [comment]
@bkunda We already have a solution for giving actions a different name in a certain menu. So we can just change the "base name" of the action that will appear in the list with shortcuts, and then override the name in menus where necessary.

**Comment by bkunda** [comment]
@MarcSabatella just a couple of quick thoughts I have following-on from your comments:
1. There might be an opportunity here to rethink the insert commands: It's extremely confusing UX to have these two insert commands that operate in different ways, while we also have different modes that do the same things. We can't expect users to look at the list of shortcuts and understand the difference between an "insert A" that replaces notes, and one that adds durations beyond the scope of the meter (even if they each do bear slightly different names).
2. When I describe "categories", I mean categories that the user would understand from interacting with the GUI, not categories (contexts) in the backend that describe how the same command works in different modes. The user doesn't need to explicitly understand the latter. What they _do_ need to be able to do, though, is see the shortcuts list in Preferences and not be perturbed by the endless number of rows. 

If these issues end up being beyond the scope of the GSOC project, then perhaps we may at least tackle them concurrently with it so we can try and improve the entire shortcut system for a later release.

**Comment by bkunda** [comment]
> @bkunda We already have a solution for giving actions a different name in a certain menu. So we can just change the "base name" of the action that will appear in the list with shortcuts, and then override the name in menus where necessary.

Sounds sensible to me!

**Comment by cbjeukendrup** [comment]
@MarcSabatella I think the current "Insert A" command is redundant, since the same can be achieved by entering the "Insert" note input mode and then adding notes using the regular commands. It seems a very niche feature to me (not redundant, only niche!), so if it would require one extra step of switching to the correct note input mode, that should be fine.

So I would be in favor of removing these "Insert A" commands, since they can only confuse people (especially since they are not in any menu of the top menu bar, so users won't be familiar with them when they see them in Preferences).
 
...unless you think there is a very strong reason that those commands should remain?

**Comment by MarcSabatella** [comment]
They are definitely redundant in that there *is* another way to do it.  But the individual insert commands are often more *efficient*, at least if you're adding just one or two notes, plus you don't then also have to remember to switch back to step-time.  And there would have to be keyboard shortcuts to move back forth between step time and insert modes, or else suddenly it's *way* less efficient.

This comes up quite often when entering Renaissance music or Romantic era piano music, and when creating educational materials.  So while obviously most people don't need this capability all that often, when you do need it, you typically need it a lot, so efficiency matters.  That is, if I might do this only one day every few weeks, but if I do this once in any given day, I'm probably doing it dozens of times, and often flipping back and forth between "entering" and "inserting" notes.

So I would say that if *we* remove the commands, we would need to add an efficient way of toggling between step-time and insert modes.  Ideally a single shortcut to toggle, not one for switching to insert and another for switching back to step time.  And I'd propose we use Ins, which is obviously the standard key for this anyhow.  We'd just need to remove note input mode from the contexts in which it functions to insert measures.  Right now we have a command called "Toggle 'Insert mode'" (shortcut Ctrl+I) but it doesn't actually do anything that I can tell.

**Comment by bkunda** [comment]
We obviously aren't going to touch input modes for MS4, but we _will_ tackle them sooner rather than later. 
I suspect there is a much more efficient way of achieving all these user goals without the impenetrability of the current "modes" system, or the confusing presence of multiple (apparently similar) shortcuts.

Certainly, the least confusing short-term solution would be @cbjeukendrup's suggestion to remove those "insert _x_" commands, noting that insert mode is still available. While it may cause an inefficiency in some user cases, it will result in a far easier-to-comprehend user experience for the vast majority of other cases (i.e. understanding which shortcuts allow you to enter notes). 

We would then look for a more efficient alternative to insert mode when we review _all_ of the current input modes to see if that entire system couldn't be improved.

**Comment by MarcSabatella** [comment]
To be clear, I wasn't asking for any actual change to note input mode behavior - just for a shortcut to toggle insert mode on and off.  Because otherwise, this important functionality is almost entirely keyboard-inaccessible.  Technically, it's still *possible* but it goes from a single keystroke to a laborious process of fumbling around with Tab, Space, and cursor keys to reach the toolbar, find the note input button, activate the drop down menu, find the correct mode, and activate it - and then having to repeat that entire process to return to standard step-time mode.  And generally have to do that over and over for each measure where you wish to do this (and as I said, if you do it once, probably you will do it a lot for that score).

So again, I'm not asking for any change in actual behavior anything - just for a shortcut to switch between step-time and insert mode.  This is practically essential for accessibility as well as usability in general if the insert commands are removed.

As mentioned, there is an existing command called "Toggle 'Insert mode'" - it just doesn't do anything.  If this command is fixed so that it actually does toggle between insert and step-time modes, that solves the problem that removing these commands would otherwise cause.  If the command is implemented as only a switch *to* insert mode, then we also need a shortcut to change back to step-time.  That command might exist currently but has no default shortcut so it doesn't show in the list, meaning I can't verify if the command is implemented at all.

**Comment by wizofaus** [comment]
From memory "Toggle 'insert mode'" does in fact do something, just something very unexpected regarding how insert mode works.  But there is already a command to enter insert mode ("Note input: enter insert mode") in v3 that can be assigned to a shortcut key, likewise to enter regular step mode ("Note input: enter step-time mode" - Alt+Shift+S by default). v4 certainly should still be able to support assigning shortcut keys to both of those actions even if the UI doesn't currently allow it.

**Comment by bkunda** [comment]
OK, first-up:

- "Toggle 'insert mode'" does indeed do something: it seems that "insert mode" actually comprises _two_ modes (or states), one that respects the meter, and one that doesn't. This shortcut toggles between these two states.

I think the path of least resistance here is to keep the two pairs of existing "enter/insert" commands, thereby avoiding the need to build any new shortcut functionality or change anything about how insert mode works. This may all fall within the remit of a more comprehensive re-design of our various note input modes in future. 

In order to solve the immediate problem of discoverability viz. the existing pairs of commands, I'd propose a re-label to (hopefully) make it clear to the user which command does what. This is my suggestion:

| Current MS4 command name | Proposed new command name |
| ----------------------------- | ------------------------------- |
| A                                                | Enter note A                                |
| Insert A                                     | Enter note A (insert mode only) |

My thinking is, if the user sees the "insert mode only" command and doesn't "get it", then they'll know it's not for them and resort to using the other one instead. Our previous research suggests that only ~12% of users make use of insert mode.  

And I'm using the verb "enter" instead of "insert" here to avoid any confusion with the actual "insert" mode. As a bonus, it will be familiar to users of MS3 as well. 

And yes, the labels are long, but we have plenty of space in the UI to accommodate even extreme localisation: 

<img width="1108" alt="Screenshot 2022-05-30 at 7 49 11 pm" src="https://user-images.githubusercontent.com/86290556/170966721-19604651-e53c-437b-9e50-433d02aad9a5.png">

This solution really just retains all the existing functionality while (hopefully) improving the discoverability of the actual command actions themselves. 

So long as this doesn't meet with any fervent objections, it would be good to lock this down so @cbjeukendrup can get back to his work 🙂

**Comment by bkunda** [comment]
Just updated the issue description as I've now separated-out the tasks of implementing the new naming convention and setting-up the categories.

**Comment by MarcSabatella** [comment]
I think this is reasonable.  Following is further discussion you can freely ignore :-)

Most users never open the shortcut customization dialog and hence don't even have the opportunity to be confused by the bad naming, but "Insert A" is asking for trouble, so I do agree short term, just renaming the command is the way to go.  I might quibble a little with the proposed new name, as the word "only" is inaccurate.  It suggests the command works in insert mode, but that's the not it at all - the whole point of the command is that it *doesn't* require the use of insert mode.  But finding a way to say that concisely is another matter.  Best I can do is "Insert note A (increases measure duration)".  If you like it or see a better variation, great, otherwise, I have no objections to "Enter note A (insert mode only)".

FWIW, I'm surprised as many as 12% of users ever use the insert functionality, and I wonder how many of them do so by accident?  It's pretty common we get a question on the forum along the lines of, "how come my measures keep getting longer and longer when I enter notes", and it turns out it's because they enable insert *mode* (not the insert *commands*, which are *much* harder to find).

Anyhow, for 4.x, I'm very happy to see us clean up how note input modes work overall.

**Comment by wizofaus** [comment]
I'd agree "Insert note A (increase measure duration)" is better - "Enter note A (insert mode only)" seems a pretty confusing way of describing it, as Marc points out, it's precisely when you're *not* in insert-mode that it's most useful.  But it doesn't seem like something that's likely to be cause a big issue with most users.

**Comment by wizofaus** [comment]
Another point re that spreadsheet, technically every possible action/command has *two* descriptions, a short one that used for the menu text (internally this is the "title") and I believe for the shortcuts dialog, and a longer one that's used for tooltips (this is the "description"), which are only shown for commands that happen to exist on toolbars. For instance if you hover the mouse over the "Save"/floppy-disk icon it says "Save score to file (Ctrl+S)", but in the Shortcuts dialog it says "File > Save" (I'm not sure how that "File > " gets added, as internally the title is just "Save", and it's not in the translation either, will investigate further).  But from a developer/programmer point of view (and this is also important for anyone writing a plugin and using the "cmd" function, there's an internal name, which is "file-save" in this case), so ideally the spreadsheet would have columns for these properties too, where they can be determined.

**Comment by MarcSabatella** [comment]
Works for me.  And we can always rationalize the phrase "insert mode only" by saying "well, it only inserts the note, it doesn't also replace what was there previously like the normal enter note commands do".

However, I'm confused by your statement that there is some sort of case in which the command does two different things depending on something or other (not clear what).  That is not true for me, nor is it supposed to be true.  For me, the insert command always inserts, period.  Always enters the note, always shifts the rest of the measure to the right, always adds time to the measure.  It acts exactly the same way under all conditions.  There shouldn't be anything inconsistent about it.  Which is why the wording I proposed is - or should be - 100% accurate: it enters the note and adds time to the measure, always.

So, are you saying there is some case in which you are seeing it *not* do this?  That shouldn't be.  In particular, when you speak of it sometimes acting to "AND not add length to a bar (all the while still pushing later notation forward in time, which is only something you'd discover if you learned it from the documentation or stumbled across it by accident" - I think whatever you stumbled on was not meant to happen at all.  If you are seeing that under some condition, that's probably just a bug.

**Comment by MarcSabatella** [comment]
Aha, now I see why there has been such confusion - wer're talking about two *entirely* different things here!  And the thing you're talking about, I didn't even knew existed (and I'm guessing no one else does, either).

What you are showing here is actually *not* insert mode, at least, not the usual insert mode that you can access from the toolbar.  What you are using is using some sort of very strange and mysterious special version of insert mode version you get with Cmd+I.  I had said that command didn't appear to work at all for me, but now I see, it does do *something*, it's just something quite different from the regular insert mode that you get using the toolbar.

So, this strange mysterious special version of insert mode does indeed alter the behavior of the "insert note" commands, to do something that seems quite useful actually, now that I know this is even possible: it inserts notes and pushes things to the right without adding beats.  To my knowledge, this is *not* something we have ever documented or supported.  It seems unlikely to be completely unintended - the behavior is too specific and too potentially useful to have been an accident.  But it's definitely *not* what insert mode actually is.

To see the *real* insert mode, you use the dropdown in the note input button on toolbar.  You'll see the icon change - which is how you know you're in real insert mode - and everything works as I've been saying - merely entering a note automatically adds time to the measure, exactly the as the insert note commands do in both this mode and in normal mode.

So, perhaps this clearly up *some* of the confusion, while now introducing new confusion- how did this strange mysterious special  insert mode get there, do we want to actually document and support it, and what does that mean in terms of this whole discussion?

**Comment by bkunda** [comment]
> To see the _real_ insert mode, you use the dropdown in the note input button on toolbar. You'll see the icon change - which is how you know you're in real insert mode - and everything works as I've been saying - merely entering a note automatically adds time to the measure, exactly the as the insert note commands do in both this mode and in normal mode.

...except when you "toggle insert mode"

https://user-images.githubusercontent.com/86290556/171210680-09d17d47-0bd5-4a73-a76e-81fa6d87921a.mov

Note here that the "insert mode" button in the note input toolbar remains active after I press `Cmd`+`I`, allowing me to enter notes _in insert mode_ without adding time to the bar.

**Comment by MarcSabatella** [comment]
Right.  I think it safe to say that since this command is entirely undocumented, probably no one has any idea that it exists or that this is intended behavior (if indeed it is).  Also, I've realized that using this is extremely likely to corrupt your score, and/or crash the program.

So, it's almost certainly out of the question to consider actually supporting this "double secret insert mode" for 4.0.  Better to make that shortcut do what it really should have done all along - toggle the *real* insert mode mode, the one that is a) shown on the toolbar, b) documented, and c) actually works :-)

**Comment by MarcSabatella** [comment]
At that point, it comes back to what I said initially - it's marginally more efficient in some case to have the individual insert commands, but with a working keyboard toggle for the real insert mode, then we could indeed remove them with relatively little harm.  In any case, definitely worth a separate issue, so I'll file one.

**Comment by bkunda** [comment]
Fantastic. Let's remove them then.
Please do file a separate issue.
@cbjeukendrup FYI!

**Comment by bkunda** [comment]
Hey @cbjeukendrup I have updated the spreadsheet (see issue description above). 
There may be more when #11586 gets merged, but this is at least a starting point. 
Let me know how you get with this. We definitely want to get it in for alpha2 before the string freeze sets in. Hopefully it's not too much work, but do let me know if you hit any snags that we need to work through 🙂 
Many thanks!

**Comment by Tantacrul** [comment]
Hey @cbjeukendrup - is this one done?

**Comment by cbjeukendrup** [comment]
No, @VanSHOE, @bkunda and me are still working on this. Problem is that the new names were based on the "few" shortcuts that were visible in Preferences, but there are a few hundreds more of them (#11060), so some names have to be changed and a lot have to be added. But I believe that will be the most time-consuming part; putting the new names into the code should be easier since I believe @VanSHOE already has a nice python script that reads the spreadsheet and generates the code.

**Comment by Tantacrul** [comment]
So, this has been targeted at alpha. My worry is strings. Will the remaining work result in large string changes?

**Comment by bkunda** [comment]
In short, I suspect that yes, it will. 
I'm now thinking this might be one of those things that just can't be done in time, particularly if we want to get translations in place sooner rather than later. I'd also like to not rush this process of naming and categorising all the new shortcuts. This may need to be bumped down.

**Comment by MarcSabatella** [comment]
Is it feasible to make this an exception to the string freeze, and clearly specify for translators to not worry about this set of strings until after this is implemented - perhaps by beta?

I suspect that would be seen as preferable to translating them all now and then re-translating them at at 4.x.  At least, if it's possible to easily identify which are the relevant strings, like if they all have the same "context" (?) in Transifex.

**Comment by cbjeukendrup** [comment]
That should be possible. All those actions have `"action"` as their context, so they are certainly distinguishable.

**Comment by bkunda** [comment]
This sounds like a sensible suggestion to me!

**Comment by Eism** [comment]
@bkunda @cbjeukendrup 
What else needs to be done to fix this issue?

**Comment by bkunda** [comment]
@eism I think we are still waiting on the PR from @VanSHOE with the corrected shortcut action titles to be merged (although I thought it had already been merged...🤔). I can't seem to locate which PR it is anymore. 

At any rate, the actions in the latest nightly currently bear the incorrect titles: 

<img width="880" alt="Screenshot 2022-07-26 at 5 44 21 pm" src="https://user-images.githubusercontent.com/86290556/180951750-48c155c2-20a1-4479-a5cd-4d9fd3a7e1ff.png">

**Comment by Eism** [comment]
@bkunda Nightly build for macOS failed. Please check on Windows.

**Comment by bkunda** [comment]
OK checked on Windows at it all looks good 👍🏻.
This issue was only to address the matter of the naming convention, which has now been done. 
When we know it works on all other platforms, I'd say this is good to close.

**Comment by Eism** [comment]
@bkunda 
MacOS, master
https://github.com/musescore/MuseScore/actions/runs/2737903132

If all is well, please close the issue

**Comment by bkunda** [comment]
@Eism shortcuts are still incorrect in MacOS

**Comment by Eism** [comment]
@bkunda need more info
Are you sure you're checking the correct build?

**Comment by bkunda** [comment]
My apologies @Eism, I had indeed opened the wrong build.
Confirming the correct action titles also appear in MacOS. Closing this issue now.
